api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java

```diff
@@ -52,6 +52,84 @@ private ProjectionUtil() {}
     }
   }
 
+  static <T> UnboundPredicate<T> truncateIntegerStrict(
+      String name, BoundPredicate<Integer> pred, Transform<Integer, T> transform) {
+    int boundary = pred.literal().value();
+    switch (pred.op()) {
+      case LT:
+        // Checking if the literal is at the lower partition boundary
+        if (transform.apply(boundary - 1).equals(transform.apply(boundary))) {
+          return predicate(Expression.Operation.LT, name, transform.apply(boundary - 1));
```

Comment:
It would also be helpful to have comments for reach case and possibly an example in the comment on line 60.

Also, I think it is easier to follow what is happening if both cases produce the same predicate operation (e.g. `LT_EQ`), and if you adjust partition values instead of adjusting the original predicate's literal.

Translating for LT produces this:

```java
int lit = pred.literal().value();
...
case LT:
  if (transform.apply(lit - 1).equals(transform.apply(lit)) {
    // lit is not the lowest value that produces the same partition value
    return predicate(LT_EQ, name, transform.apply(lit) - 1);
  } else {
    // lit is the lowest value that produces the partition value
    return predicate(LT_EQ, name, transform.apply(lit) - 1);
  }
```

As you can see, the result is the same so we can get rid of the check in this case. And that makes sense: `date(2019-05-13T00:00:00)` is `2019-05-13`, so the strict projection for `<` should be `<= 2019-05-13`, the same as for `2019-05-13T01:01:01`.

I think you only need the boundary check for `LT_EQ` and `GT_EQ`, where the literal may be the highest value that produces the partition value.

Reply:
Ack, will add comments and examples. 

I had originally started with keeping the operation same to `LT_EQ`, but then that makes it specific to only Integer values. The same logic might be applicable to the truncate ones too? So did not want to tie it to a type.

Reply:
Do you still think it should be Integer specific?

Reply:
Actually just checked, right now it's only used by truncate integers and date/timestamps, so should be fine making it Integer

Reply:
truncateLong is getting used for both Timestamps and Truncate.LONG